clang/lib/Parse/ParseExpr.cpp

```diff
@@ -1824,6 +1833,61 @@ ExprResult Parser::ParseCastExpression(CastParseKind ParseKind,
   return Res;
 }
 
+/// __builtin_num_fields '(' type-id ')' or
+/// __builtin_num_bases '(' type-id ')'
+ExprResult Parser::ParseSYCLBuiltinNum(bool IsNumFields) {
+  assert((IsNumFields && Tok.is(tok::kw___builtin_num_fields)) ||
```

Comment:
We could do that, but I think it's somewhat more confusing of a parsing interface because it's no longer explicit *what* you expect to parse or why you're parsing it. That said, I don't feel strongly and you did get confused (twice!), so I'll change it.